cranelift/filetests/filetests/isa/aarch64/amodes.clif

```diff
@@ -72,18 +72,20 @@ block0(v0: i64, v1: i32):
 
 ; VCode:
 ; block0:
-;   add x3, x0, #68
-;   add x5, x3, x0
-;   add x7, x5, x1, SXTW
-;   ldr w0, [x7, w1, SXTW]
+;   sxtw x7, w1
+;   add x7, x7, #32
+;   add x7, x7, x0
+;   add x6, x7, #4
```

Comment:
This is a little odd -- we no longer fold the sign-extend into its two uses with builtin extend modes on the `add` and in the `AMode`; and also we have two separate adds of constants that don't get folded together. Any idea what's causing this?

Reply:
For the constants I know those aren't combined because one is an `iconst` operand to an add and one is the offset listed on the `load` instruction. Those aren't currently combined with egraphs (or anywhere else I think) so it ends up coming out here as two constants. Looks like wasm translation doesn't use the offset on `load` and `store` though so this at least shouldn't affect that.

The sign-extend I think doens't pattern-match here due to its depth in the `iadd` tree which looks like:

```
                +
             -------
            /        \
           +          +
          / \        / \
         +  v0      +  v0
        / \        / \
       /   32     /   32
      /          /
  sext(v1)   sext(v1)
```

there the `sextend` nodes are so deep in the tree the pattern matching isn't picking them out.

That's at least why it's not matching, but as to the impact of it I'm not really sure. I don't think this affects wasm much since wasm typically has one main sign-extend to pattern match, but I'm also not super-certain about that.

Reply:
OK, yeah, that makes sense. The old handwritten code would traverse the tree arbitrarily deep and gather all the constant offsets, extends, etc., which is why it can do better here I think. But it's possible that this doesn't matter too much in practice.

If we can just spot-check performance (`bz2` and `spidermonkey` from Sightglass or whatnot) to make sure, I think I'm happy with this regression at the fringe.

Reply:
Sightglass shows:


```

execution :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  Δ = 142380.50 ± 5778.82 (confidence = 99%)

  new-amode.so is 1.02x to 1.02x faster than main.so!

  [8232969 8267115.57 8308418] main.so
  [8072036 8124735.07 8159535] new-amode.so

compilation :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  Δ = 1706947.06 ± 1093497.57 (confidence = 99%)

  new-amode.so is 1.00x to 1.02x faster than main.so!

  [151473126 155506902.82 158095691] main.so
  [145110211 153799955.76 160548952] new-amode.so
```

and everything else is "No difference in performance."

I haven't done much to make this machine super reliable about performance gathering, so I'd interpret this as "no difference" across the board and the 2% here is probably noise.

Reply:
Sounds good -- ship it!